### ISSUE TITLE
Rule Proposal: require-code-point-escape

### ISSUE THREAD

**Issue by kufii** [eslint/eslint#12488]
This proposal is to require the new ES6 code point escapes eg. `'\u{1F4A9}'`

Examples of **incorrect** code for this proposal:

```javascript
let foo = "\123"; // octal
let foo = "\x7A"; // hex
let foo = "\u2661"; // unicode escape sequence
let foo = "\uD83D\uDCA9"; // unicode surrogate pair
```

Examples of **correct** code for this proposal:

```javascript
let foo = "\u{7A}";
let foo = "\u{1F4A9}";
```

**Comment by kaicataldo** [comment]
@kufii Can you elaborate on what problem you're trying to solve with this proposal?

**Comment by kufii** [comment]
I'm specifically concerned about the usage of surrogate pair unicode escape sequences (eg. `\uD83D\uDCA9`, which actually points to hex value `1F4A9`), as they are confusing since neither of the hex codes point to the actual unicode hex value, and instead are used to calculate the derived hex value.

But I figured if you're disallowing those you might as well disallow all the others too.

The new unicode code point escapes are much clearer since they always point to the exact hex code value.

So in summary `\u{1F4A9}` > `\uD83D\uDCA9`
